File: services/email-ingestion/src/interfaces/email_service.py
"""
EmailService Interface

Provider-agnostic interface for email services.
Enables swapping providers (Gmail API, IMAP, SendGrid, etc.) without refactoring.
"""

from abc import ABC, abstractmethod
from typing import List, Optional
from datetime import datetime

from ..types.email import (
    Email,
    EmailAttachment,
    Inbox,
    ProviderCapabilities,
    WatchHandle,
)


class EmailService(ABC):
    """
    Abstract interface for email services.
    
    All email providers must implement this interface.
    """
    
    @abstractmethod
    def fetchEmails(self, inbox: str, since: Optional[datetime] = None) -> List[Email]:
        """
        Fetch emails from inbox.
        
        Args:
            inbox: Inbox address or identifier
            since: Optional date to fetch emails since
            
        Returns:
            Array of emails
        """
        pass
    
    @abstractmethod
    def getEmailById(self, emailId: str) -> Optional[Email]:
        """
        Get email by provider-specific ID.
        
        Args:
            emailId: Provider-specific email ID
            
        Returns:
            Email or None if not found
        """
        pass
    
    @abstractmethod
    def markAsProcessed(self, emailId: str, label: Optional[str] = None) -> None:
        """
        Mark email as processed (label, read, etc.).
        
        Args:
            emailId: Provider-specific email ID
            label: Optional label/folder name
        """
        pass
    
    @abstractmethod
    def sendEmail(
        self,
        to: str,
        subject: str,
        body: str,
        htmlBody: Optional[str] = None,
        inReplyTo: Optional[str] = None
    ) -> None:
        """
        Send email.
        
        Args:
            to: Recipient email address
            subject: Email subject
            body: Plain text body
            htmlBody: Optional HTML body
            inReplyTo: Optional email ID to reply to
        """
        pass
    
    @abstractmethod
    def getAttachment(self, emailId: str, attachmentId: str) -> bytes:
        """
        Get email attachment.
        
        Args:
            emailId: Provider-specific email ID
            attachmentId: Provider-specific attachment ID
            
        Returns:
            Attachment data as bytes
        """
        pass
    
    @abstractmethod
    def listInboxes(self) -> List[Inbox]:
        """
        List available inboxes.
        
        Returns:
            Array of inboxes
        """
        pass
    
    @abstractmethod
    def watchInbox(
        self,
        inbox: str,
        callback: callable
    ) -> WatchHandle:
        """
        Watch inbox for new emails (webhooks, push notifications).
        
        Args:
            inbox: Inbox address or identifier
            callback: Callback function for new emails
            
        Returns:
            Watch handle to stop watching
        """
        pass
    
    @abstractmethod
    def getProviderName(self) -> str:
        """
        Get provider name (e.g., "gmail", "imap", "sendgrid").
        
        Returns:
            Provider name
        """
        pass
    
    @abstractmethod
    def getProviderCapabilities(self) -> ProviderCapabilities:
        """
        Get provider capabilities.
        
        Returns:
            Provider capabilities
        """
        pass

File: services/email-response-generation/src/interfaces/email_service.ts
/**
 * EmailService Interface
 * 
 * Provider-agnostic interface for email services.
 * Enables swapping providers (Gmail API, IMAP, SendGrid, etc.) without refactoring.
 */

import type {
  Email,
  EmailAttachment,
  Inbox,
  ProviderCapabilities,
  WatchHandle,
} from '../types/email';

/**
 * EmailService interface
 * 
 * All email providers must implement this interface.
 */
export interface EmailService {
  /**
   * Fetch emails from inbox
   * 
   * @param inbox - Inbox address or identifier
   * @param since - Optional date to fetch emails since
   * @returns Array of emails
   */
  fetchEmails(inbox: string, since?: Date): Promise<Email[]>;

  /**
   * Get email by provider-specific ID
   * 
   * @param emailId - Provider-specific email ID
   * @returns Email or null if not found
   */
  getEmailById(emailId: string): Promise<Email | null>;

  /**
   * Mark email as processed (label, read, etc.)
   * 
   * @param emailId - Provider-specific email ID
   * @param label - Optional label/folder name
   */
  markAsProcessed(emailId: string, label?: string): Promise<void>;

  /**
   * Send email
   * 
   * @param to - Recipient email address
   * @param subject - Email subject
   * @param body - Plain text body
   * @param htmlBody - Optional HTML body
   * @param inReplyTo - Optional email ID to reply to
   */
  sendEmail(
    to: string,
    subject: string,
    body: string,
    htmlBody?: string,
    inReplyTo?: string
  ): Promise<void>;

  /**
   * Get email attachment
   * 
   * @param emailId - Provider-specific email ID
   * @param attachmentId - Provider-specific attachment ID
   * @returns Attachment data as Buffer
   */
  getAttachment(emailId: string, attachmentId: string): Promise<Buffer>;

  /**
   * List available inboxes
   * 
   * @returns Array of inboxes
   */
  listInboxes(): Promise<Inbox[]>;

  /**
   * Watch inbox for new emails (webhooks, push notifications)
   * 
   * @param inbox - Inbox address or identifier
   * @param callback - Callback function for new emails
   * @returns Watch handle to stop watching
   */
  watchInbox(inbox: string, callback: (email: Email) => void): Promise<WatchHandle>;

  /**
   * Get provider name
   * 
   * @returns Provider name (e.g., "gmail", "imap", "sendgrid")
   */
  getProviderName(): string;

  /**
   * Get provider capabilities
   * 
   * @returns Provider capabilities
   */
  getProviderCapabilities(): ProviderCapabilities;
}
